st2common/st2common/util/auth.py

```diff
@@ -58,3 +66,55 @@ def validate_token(token_in_headers, token_in_query_params):
 
     LOG.audit('Token with id "%s" is validated.' % (token.id))
     return token
+
+
+def generate_api_key():
+    """
+    Generates an sufficiently large and random key.
+
+    credit: http://jetfar.com/simple-api-key-generation-in-python/
+    """
+    # 256bit seed from urandom
+    seed = os.urandom(256)
+    # since urandom does not provide sufficient entropy hash, base64encode and salt.
+    # The resulting value is now large and should be hard to predict.
+    hashed_seed = hashlib.sha256(seed).hexdigest()
+    return base64.b64encode(
+        hashed_seed,
+        random.choice(['rA', 'aZ', 'gQ', 'hH', 'hG', 'aR', 'DD'])).rstrip('==')
+
+
+def generate_api_key_and_hash():
+    api_key = generate_api_key()
+    api_key_hash = hash_utils.hash(api_key)
+    return api_key, api_key_hash
+
+
+def validate_api_key(api_key_in_headers, api_key_query_params):
+    """
+    Validate the provided API key.
+
+    :param api_key_in_headers: API key provided via headers.
+    :type api_key_in_headers: ``str``
+
+    :param api_key_query_params: API key provided via query params.
+    :type api_key_query_params: ``str``
+
+    :return: TokenDB object on success.
+    :rtype: :class:`.ApiKeyDB`
+    """
+    if not api_key_in_headers and not api_key_query_params:
+        LOG.audit('API key is not found in header or query parameters.')
+        raise exceptions.ApiKeyNotProvidedError('API key is not provided.')
+
+    if api_key_in_headers:
+        LOG.audit('API key provided in headers')
+
+    if api_key_query_params:
+        LOG.audit('API key provided in query parameters')
+
+    api_key = api_key_in_headers or api_key_query_params
```

Comment:
I just grepped for hash_utils and I found it gets hashed in the model get method - https://github.com/StackStorm/st2/pull/1964/files#diff-a15ecbc227b545b35658d0cbb697c99dR73.

I'm personally not a big fan of that since the DB model itself doesn't and shouldn't know anything about the raw key and how to handle that. I would prefer to hash it either inside the auth hook or here.